nixpkgs/toolchains/rust.bzl

```
@@ -0,0 +1,159 @@
+                    "lib/rustlib/*/lib/*.so",
+                    "lib/rustlib/*/lib/*.dylib",
+                    "lib/rustlib/*/lib/*.a",
+                    "lib/rustlib/*/lib/self-contained/**",
+                ],
+                # Some patterns (e.g. `lib/*.a`) don't match anything, see https://github.com/bazelbuild/rules_rust/pull/245
+                allow_empty = True,
+            ),
+            visibility = ["//visibility:public"],
+        )
+
+        filegroup(
+            name = "rust_doc",
+            srcs = ["bin/rustdoc"],
+            visibility = ["//visibility:public"],
+        )
+
+        load('@rules_rust//rust:toolchain.bzl', 'rust_toolchain')
+        rust_toolchain(
+            name = "rust_nix_impl",
+            rustc = ":rustc",
+            rustc_lib = ":rustc_lib",
+            rust_lib = ":rust_lib",
+            rust_doc = ":rust_doc",
+            binary_ext = "",
+            staticlib_ext = ".a",
+            dylib_ext = ".so",
+            stdlib_linkflags = ["-lpthread", "-ldl"],
+            os = "${os}",
+            target_triple = "${target}",
+        )
+        EOF
+    '';
+}
+"""
+
+_rust_nix_toolchain = """
+toolchain(
+    name = "rust_nix",
+    toolchain = "@{toolchain_repo}//:rust_nix_impl",
+    toolchain_type = "@rules_rust//rust:toolchain",
+    exec_compatible_with = {exec_constraints},
+    target_compatible_with = {target_constraints},
+)
+"""
+
+def _ensure_constraints(repository_ctx):
+    cpu = get_cpu_value(repository_ctx)
+    os = {"darwin": "osx"}.get(cpu, "linux")
+    if not repository_ctx.attr.target_constraints and not repository_ctx.attr.exec_constraints:
+        target_constraints = ["@platforms//cpu:x86_64"]
+        target_constraints.append("@platforms//os:{}".format(os))
+        exec_constraints = target_constraints
+    else:
+        target_constraints = list(repository_ctx.attr.target_constraints)
+        exec_constraints = list(repository_ctx.attr.exec_constraints)
+    exec_constraints.append("@io_tweag_rules_nixpkgs//nixpkgs/constraints:support_nix")
+    return exec_constraints, target_constraints
+
+def _nixpkgs_rust_toolchain_impl(repository_ctx):
+    exec_constraints = [ "@io_tweag_rules_nixpkgs//nixpkgs/constraints:support_nix", "@platforms//cpu:x86_64" ]
+    target_constraints = [ "@platforms//cpu:x86_64" ]
+    cpu = get_cpu_value(repository_ctx)
+    repository_ctx.file(
+            "BUILD.bazel",
+            executable = False,
+            content = _rust_nix_toolchain.format(
+                    toolchain_repo = repository_ctx.attr.toolchain_repo,
+                    exec_constraints = exec_constraints,
+                    target_constraints = target_constraints
+            )
+    )
+
+_nixpkgs_rust_toolchain = repository_rule(
+    _nixpkgs_rust_toolchain_impl,
+    attrs = {
+        "toolchain_repo": attr.string(),
+    },
+)
+
+def nixpkgs_rust_configure(
+        sdk_name = "rust_linux_x86_64",
```

```suggestion
        name = "nixpkgs_rust_linux_x86_64",
```

Using `name` for consistency with the other rules. Go is the only counter example where `sdk_name` is used for consistency with the Go rules. But, rules_rust doesn't seem to have that convention.

Using `nixpkgs_` prefix to reduce likelihood of collision with a non-Nix repository.

---

The collision was actually intended: cargo_raze has a hard-coded dependency on `rust_{os}_{arch}//:cargo`, so the easiest way to use it was to register the Nix-provided toolchain with that name. Of course, this can also be user-specified, and as it is cargo_raze specific, I agree that it's probably best to default to a distinct name.